plugins/outputs/zabbix/zabbix_test.go

```diff
@@ -0,0 +1,868 @@
+					require.Zero(t, len(test.zabbixMetrics), "no metrics should be expected if the connection times out")
+				}
+
+				wg.Done()
+			}()
+
+			err = z.Write(test.telegrafMetrics)
+			require.NoError(t, err)
+
+			// Wait for zabbix server emulator to finish
+			wg.Wait()
+		})
+	}
+}
+
+// TestLLD tests how LLD metrics are sent simulating the time passing.
+// LLD is sent each LLDSendInterval. Only new data.
+// LLD data is cleared LLDClearInterval.
+func TestLLD(t *testing.T) {
+	z := &Zabbix{
+		Address:          "127.0.0.1:10051",
+		Prefix:           "telegraf.",
+		HostTag:          "host",
+		LLDSendInterval:  config.Duration(10 * time.Minute),
+		LLDClearInterval: config.Duration(1 * time.Hour),
+		Log:              testutil.Logger{},
+	}
+	err := z.Init()
+	require.NoError(t, err)
+
+	// Telegraf metric which will be sent repeatedly
+	m := testutil.MustMetric(
+		"name",
+		map[string]string{"host": "hostA", "foo": "bar"},
+		map[string]interface{}{"value": int64(0)},
+		time.Unix(0, 0),
+	)
+
+	mNew := testutil.MustMetric(
+		"name",
+		map[string]string{"host": "hostA", "foo": "moo"},
+		map[string]interface{}{"value": int64(0)},
+		time.Unix(0, 0),
+	)
+
+	// Expected Zabbix metric generated
+	zabbixMetric := zabbixRequestData{
+		Host:  "hostA",
+		Key:   "telegraf.name.value[bar]",
+		Value: "0",
+		Clock: 0,
+	}
+
+	// Expected Zabbix metric generated
+	zabbixMetricNew := zabbixRequestData{
+		Host:  "hostA",
+		Key:   "telegraf.name.value[moo]",
+		Value: "0",
+		Clock: 0,
+	}
+
+	// Expected Zabbix LLD metric generated
+	zabbixLLDMetric := zabbixRequestData{
+		Host:  "hostA",
+		Key:   "telegraf.lld.name.foo",
+		Value: `{"data":[{"{#FOO}":"bar"}]}`,
+		Clock: 0,
+	}
+
+	// Expected Zabbix LLD metric generated
+	zabbixLLDMetricNew := zabbixRequestData{
+		Host:  "hostA",
+		Key:   "telegraf.lld.name.foo",
+		Value: `{"data":[{"{#FOO}":"bar"},{"{#FOO}":"moo"}]}`,
+		Clock: 0,
+	}
+
+	// Simulate a Zabbix server to get the data sent
+	listener, lerr := net.Listen("tcp", z.Address)
+	require.NoError(t, lerr)
+
+	defer listener.Close()
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+
+	// Read first packet with two metrics, then the first autoregister packet and the second autoregister packet.
+	go func() {
+		// First packet with metrics
+		request := listenForZabbixMetric(t, listener, false)
+		compareData(t, []zabbixRequestData{zabbixMetric}, request.Data)
+
+		// Second packet, while time has not surpassed LLDSendInterval
+		request = listenForZabbixMetric(t, listener, false)
+		compareData(t, []zabbixRequestData{zabbixMetric}, request.Data)
+
+		// Third packet, time has surpassed LLDSendInterval, metrics + LLD
+		request = listenForZabbixMetric(t, listener, false)
+		if len(request.Data) == 2 {
+			request.Data[1].Clock = 0 // Ignore lld request clock
+			compareData(t, []zabbixRequestData{zabbixMetric, zabbixLLDMetric}, request.Data)
+		} else {
+			t.Errorf("Expected 2 metrics, got %d", len(request.Data))
+		}
```

Comment:
```suggestion
		request = listenForZabbixMetric(t, listener, false)
		require.Len(t, request.Data, 2, "Expected 2 metrics")
		request.Data[1].Clock = 0 // Ignore lld request clock
		compareData(t, []zabbixRequestData{zabbixMetric, zabbixLLDMetric}, request.Data)
```
Or maybe
```suggestion
		request = listenForZabbixMetric(t, listener, false)
		if assert.Len(t, request.Data, 2, "Expected 2 metrics") {
			request.Data[1].Clock = 0 // Ignore lld request clock
			compareData(t, []zabbixRequestData{zabbixMetric, zabbixLLDMetric}, request.Data)
		}
```
And same for following occurrences.